### PR TITLE
Fix issue #127: new rules for uses inside macros

### PR DESCRIPTION
**Pull request by kimgr**
This isn't 100% finished, but I wanted to show it off and get some feedback. This patch as-is *should* fix issue 127, but I'd like to expand tests a little bit to cover a few corner cases.

What do you think of the overall approach? I'm not all that happy with the details (the pair returned from GetUseLocationForMacroExpansion, for example), but I think the general direction is right.

See commit comment for more details.

### CODE REVIEW

**Comment by vsapsai** [comment]
Can you please describe in general what went wrong in issue #127? I see you are changing how we attribute decls in macro, to caller or to macro author. But I am missing why are you doing this. A small test case that demonstrates the problem will be very useful.

**Comment by kimgr** [comment]
The original problem as reported was a construct like this:

    // util.h
    class Util {
    public:
      void Foo() {
         // Interesting stuff here, I'm sure.
      }
    };

    // macros.h
    #define DECLARE_UTIL(CLASSNAME) \
        static class Util ms_util; \
        void Foo();

    #define DEFINE_UTIL(CLASSNAME) \
        Util CLASSNAME::ms_util; \
        void CLASSNAME::Foo() { \
          ms_util.Foo();  \
        }

(this is cut down from the attachment from issue #127 on Google Code.)

The original problem was that they wanted to push the responsibility for `util.h` onto clients. This makes sense because using `DECLARE_UTIL` does not require `util.h` to be included, but using `DEFINE_UTIL` does, and it seems intuitive to keep the macros together in the same file.

So I suggested the same heuristic for author intent that we use for templates and typedefs: if the macro file forward-declares a type, responsibility for that type should be pushed onto files instantiating the macro.

In trying to understand IWYU's current handling of macro locations I thought it committed to the location too early (in `CurrentLoc`), so I delayed that decision until we were ready to report the use. That way I could include any forward-declares in the analysis, but I also had to re-work the handling of macro arguments (which worked magically before.)

I'm now pretty convinced that the stuff I'm doing with macro arguments is wrong, and I'd like to explore that further, but hopefully this sheds some light on the background. Any ideas on how to document this better in code would be welcome.

**Comment by vsapsai** [comment]
Delaying macro location from `CurrentLoc` to `GetUseLocationForMacroExpansion` seems to be fine. Not sure about returning `isForwardDeclaredInMacroFile` as the second part of a pair. It is set only in one place inside `GetUseLocationForMacroExpansion`, though it has multiple branches. And it is used only in one place. So I think it is cleaner not to return a pair. The only benefit I see is unknown performance improvement. `GetUseLocationForMacroExpansion` will calculate `isForwardDeclaredInMacroFile` anyway so if it is too heavy, recalculating it in other places can be too expensive.

Is all the magic with macro arguments finished and ready for review? Because all this `FindNextIdentifier` looks tricky.

**Comment by kimgr** [comment]
Yes, the reason I returned the pair was because it seemed unnecessary to calculate the same thing more than once, but it does muddy up the logic, so I'll try to kill that darling to get something that's easier to understand.

No, I think I'm going about the macro arguments the wrong way. Here's an example where it will break, for example:

    // macros.h
    #define APPLY(a, op, b) \
        a op b

    // fiasco.h
    int operator+(int a, int b) {
      return a * b;
    }

    // main.cc
    #include "macros.h"
    #include "fiasco.h"

    void foo() {
       int c = APPLY(100, +, 200);
    }

Because the operator comes in as a macro argument, its use should be attributed to `main.cc`. But the current code is just looking for identifiers, so it can never work.

I saw test failures with the classical array-size macro:

    #define ARRAY_SIZE(x) sizeof(x) / sizeof(*x)

Something caused the use-location for `x` to point somewhere else, that's why I started searching through the macro body with `FindNextIdentifier`. Now I hope I can adjust the use-location so it points to the right thing, instead, and thereby solve the operator-as-argument case as well.

So, long story short, no, I don't think the macro argument magic is ready for review, I think it's broken :-)

**Comment by kimgr** [comment]
Abandoning this, I have a new branch with a working approach.
